### PR TITLE
 LTS MatrixEntry support

### PR DESCRIPTION
**Pull request by kdvalin**
This PR expands out a MatrixEntry to support LTS payloads.

General structure expected:
```json5
{
   data: {
        metrics: { 
            ...
         },
         thresholds: {
             ...
          },
          ...
       },
    metadata: {
        ...
     }
}
```

### CODE REVIEW

File: matrix_benchmarking/common.py [hunk 1]
```diff
@@ -1,34 +1,76 @@
+from typing import Iterator
+
+import uuid
```
Comment (kpouget) [inline matrix_benchmarking/common.py:3]:
```suggestion
```

File: matrix_benchmarking/common.py [hunk 2]
```diff
@@ -1,34 +1,76 @@
...
+        self.is_lts = is_lts
+        self.is_gathered = False
 
-        self.location = location
-        self.results = results
+        if not is_lts:
+            if settings:
+                self.settings = settings
+            else:
+                self.settings = types.SimpleNamespace()
```
Comment (kpouget) [inline matrix_benchmarking/common.py:25]:
```suggestion
            self.settings = settings or types.SimpleNamespace()
```

File: matrix_benchmarking/common.py [hunk 3]
```diff
@@ -40,18 +82,36 @@ def settings_to_key(settings):
...
+        for settings_values in sorted(itertools.product(*setting_lists)):
+            settings.update(dict(settings_values))
+            key = Matrix.settings_to_key(settings)
+            try:
+                entry = Matrix.processed_map[key]
+            except KeyError:
+                continue # missing experiment
+            if (entry.is_lts and include_lts) or (not entry.is_lts and include_local):
+                count += 1
+        return count
```
Comment (kpouget) [inline matrix_benchmarking/common.py:116]: I think this is equivalent to:
```suggestion
        return len(self.all_records(settings, setting_lists, include_local, include_lts))
```

could you also add this:
```
    @staticmethod
    def has_records(settings, setting_lists, include_local=True, include_lts=False):
         for _ in  self.all_records(settings, setting_lists, include_local, include_lts):
              return True
         return False
```
(this one avoid unnecessary traversal of the matrix, which can be useful where there are hundreds of records)

File: matrix_benchmarking/visualize.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def run():
         logging.info(f"Loading results ... ")
         workload_store.parse_data()
         logging.info(f"Loading results ... done. Found {len(common.Matrix.processed_map)} results.")
-        if not common.Matrix.processed_map:
+        if not common.Matrix.processed_map and not common.Matrix.lts_processed_map:
```
Comment (kpouget) [inline matrix_benchmarking/visualize.py:57]:
```suggestion
        if not common.Matrix.processed_map:
```

Comment (kpouget) [reply matrix_benchmarking/visualize.py:57]: (IIUC)

**Review by kpouget — COMMENTED** [review]
thanks @kdvalin, looks good :) 
I added minor comments, we should be able to merge this one soon!

**Comment by kpouget** [comment]
@kdvalin could you squash the commits, and keep only one per file? (or more than one if it logically makes sense)

File: matrix_benchmarking/common.py [hunk 5]
```diff
@@ -1,34 +1,52 @@
...
 class MatrixEntry(types.SimpleNamespace):
     def __init__(self, location, results,
                  processed_key, import_key,
                  processed_settings, import_settings, settings=None,
-                 stats=None, is_gathered=None):
-        if settings:
-            self.settings = settings
-        else:
-            self.settings = types.SimpleNamespace()
+                 stats=None, is_gathered=None, is_lts = False):
```
Comment (kpouget) [inline matrix_benchmarking/common.py:14]:
```suggestion
                 stats=None, is_gathered=None, is_lts=False):
```

File: matrix_benchmarking/common.py [hunk 6]
```diff
@@ -40,18 +58,31 @@ def settings_to_key(settings):
...
             settings.update(dict(settings_values))
             key = Matrix.settings_to_key(settings)
+
             try:
-                yield Matrix.processed_map[key]
+                entry = Matrix.processed_map[key]
             except KeyError:
                 continue # missing experiment
+            if (entry.is_lts and include_lts) or (not entry.is_lts and include_local):
+                yield entry
```
Comment (kpouget) [inline matrix_benchmarking/common.py:71]: I wonder if one way or another, `all_records` should return the local entries first 🤔 
(see https://github.com/openshift-psap/ci-artifacts/pull/697 for why it's important)

actually, I think the sorting should be done here, at the core level.

but I'm not sure what's the best way to expose it, and what should be the default 🤔 

with the current code:
```python
for settings_values in sorted(itertools.product(*setting_lists)):
```
the order is defined by `setting_list`, which itself comes from there:
```python
def do_plot(self, ordered_vars, settings, setting_lists, variables, cfg):
```
so in the web UI (that you're not using, AFAIK), you can change the order to decide what comes first.

I guess I would prefer having another flag `local_first`, which behaves this way:
```python
def all_records(settings, setting_lists, include_local=True, include_lts=False, local_first=True) -> Iterator[MatrixEntry]:
    if local_first:
        if not (include_local and include_lts):
            raise ValueError("local_first=True cannot be set without include_local=True and include_lts=True")
        yield from all_records(settings, setting_list, include_local=True, include_lts=False, local_first=False)
        yield from all_records(settings, setting_list, include_local=False, include_lts=True, local_first=False)
        return
```

File: matrix_benchmarking/common.py [hunk 7]
```diff
@@ -1,34 +1,70 @@
...
-                 stats=None, is_gathered=None):
-        if settings:
-            self.settings = settings
+                 stats=None, is_gathered=None, is_lts = False):
+        self.is_lts = is_lts
+        self.is_gathered = False
+
+        if is_lts:
+            self.metadata = results['metadata']
+            self.data = results['data']
```
[the remark this answers is not in the document]
Comment (kdvalin) [reply matrix_benchmarking/common.py]: fixed

File: matrix_benchmarking/common.py [hunk 8]
```diff
@@ -1,34 +1,70 @@
...
-        self.processed_key = processed_key
-        self.import_settings = processed_settings
+            self.location = location
+            self.results = results
 
-        self.is_gathered = False
+            self.settings.__dict__.update(processed_settings)
+
+            self.processed_key = processed_key
+            self.import_settings = processed_settings
```
[the remark this answers is not in the document]
Comment (kdvalin) [reply matrix_benchmarking/common.py]: Should be fixed now

File: matrix_benchmarking/store/simple.py [hunk 9]
```diff
@@ -62,6 +62,7 @@ def _parse_directory(results_dir, expe, dirname):
                         continue
 
                     import_settings[key] = value
+                    import_settings['lts'] = 0
```
Comment (kpouget) [inline matrix_benchmarking/store/simple.py:65]: can you remove this?
I prefer not to have it at the moment, we'll decide later if that's mandatory or not for having a proper sorting

**Review by kpouget — COMMENTED** [review]
thanks @kdvalin , apart from the lts=0 setting, I think the PR is good for merging
I just want to see a test run showing the actual results

File: matrix_benchmarking/store/simple.py (hunk 9 above)
Comment (kpouget) [reply matrix_benchmarking/store/simple.py:65]: ping, this should be removed

**Comment by kpouget** [comment]
thanks @kdvalin , merging :)

**Pull request merged**